src/libsystemd-network/network-internal.c

```diff
@@ -272,25 +273,35 @@ int config_parse_hwaddr(const char *unit,
                         void *userdata) {
         struct ether_addr **hwaddr = data;
         struct ether_addr *n;
-        int r;
+        int r, before = 0, after = 0;
+        const char *start = rvalue, *end;
 
         assert(filename);
         assert(lvalue);
         assert(rvalue);
         assert(data);
 
+        while (isspace(*start))
+                start++;
+
         n = new0(struct ether_addr, 1);
         if (!n)
                 return log_oom();
 
-        r = sscanf(rvalue, "%02hhx:%02hhx:%02hhx:%02hhx:%02hhx:%02hhx",
+        r = sscanf(start, "%n%02hhx:%02hhx:%02hhx:%02hhx:%02hhx:%02hhx%n",
+                   &before,
                    &n->ether_addr_octet[0],
                    &n->ether_addr_octet[1],
                    &n->ether_addr_octet[2],
                    &n->ether_addr_octet[3],
                    &n->ether_addr_octet[4],
-                   &n->ether_addr_octet[5]);
-        if (r != 6) {
+                   &n->ether_addr_octet[5],
+                   &after);
+        end = start + after;
```

Comment:
thanks for pointing out that ether_ntoa drops leading octets.

ether_ntoa(3) says:

    The ether_ntoa() function converts the Ethernet host address addr given
    in  network  byte  order  to a string in standard hex-digits-and-colons
    notation, omitting leading zeros.  

so right, we need to handle that, which the series i proposed does not
do.

On Wed 2016-05-11 13:42:20 -0400, Lennart Poettering wrote:
> Hmm, I think we should probably replace sscanf() here with
> ether_aton_r(), unless there's a good reason why that parser isn't
> good enough for us. @teg, any reason why you used sscanf() here?

Running the corrected list of tests on ether_aton_r() after strspn to
find the start of the non-whitespace gives me the following failures:

```
1. 'aa:bb:cc:dd:ee:ff \t\nxxx' should not have matched (matched as: aa:bb:cc:dd:ee:ff)
2. 'aa:bb:cc:dd:ee:ff:gg' should not have matched (matched as: aa:bb:cc:dd:ee:ff)
3. 'aa-bb-cc-dd-ee-ff' failed to match (should have been: aa:bb:cc:dd:ee:ff)
4. 'AA-BB-CC-DD-EE-FF' failed to match (should have been: aa:bb:cc:dd:ee:ff)
5. '01-23-45-67-89-ab' failed to match (should have been: 1:23:45:67:89:ab)
6. 'aabb.ccdd.eeff' failed to match (should have been: aa:bb:cc:dd:ee:ff)
7. '0123.4567.89ab' failed to match (should have been: 1:23:45:67:89:ab)
```

Even if IEEE- and Cisco-style addresses aren't worth caring about,
there's still the suffix problem to consider.  A mis-typed MAC address
with an extra octet anywhere in it will be silently accepted (and the
final octet ignored, regardless of which one was mistyped).  Seems like
it would be nice to alert the user that something is amiss when it's
detectable.

        --dkg